File: Services/OrdersService.cs
using EcomDash.Models;
using EcomDash.Requests;
using EcomDash.Utils;
using System.Dynamic;

namespace EcomDash.Services
{
    public class OrdersService
    {
        EcommerceContext dbContext = new EcommerceContext();
        OrdersPaginationUtils paginationUtils = new OrdersPaginationUtils();

        public Customer getCustomerOrder(int orderId)
        {
            Customer customer = dbContext.Orders.Where(x => x.Id == orderId)
                .Select(x => x.Customer).FirstOrDefault();
            return customer;
        }

        public ExpandoObject getProductsOrder(int orderId)
        {
            List<Product> products = dbContext.OrderDetails.Where(x => x.Order.Id == orderId)
                .Select(x => x.Product).ToList();

            dynamic result = new ExpandoObject();
            result.Products = products;
            result.count = products.Count;
            return result;
        }

        public ExpandoObject getPaginatedOrders(string orderBy, string orderDirection, int currentPage = 1, int pageSize = 5)
        {
            return paginationUtils.GetPaginatedOrders(ref dbContext, orderBy, orderDirection, currentPage, pageSize);
        }

        public Order CreateOrder(OrderDto orderDto)
        {
            Customer customer = dbContext.Customers.Find(orderDto.CustomerId);

            var order = new Order
            {
                CustomerId = customer.Id,
                OrderDate = DateTime.Now,
                ShippingAddress = customer.DefaultShippingAddress,
                OrderAddress = orderDto.ShippingAddress,
                OrderEmail = customer.Email,
                OrderStatus = orderDto.OrderStatus
            };

            decimal totalAmount = 0;

            foreach (var productDto in orderDto.Products)
            {
                var product = dbContext.Products.Find(productDto.ProductId);
                if (product == null)
                {
                    throw new Exception($"Prodotto non trovato: {productDto.ProductId}");
                }

                var orderDetail = new OrderDetail
                {
                    ProductId = productDto.ProductId,
                };

                order.OrderDetails.Add(orderDetail);

                totalAmount += (decimal)(product.Price * Convert.ToDecimal(productDto.Quantity));
            }

            order.Amount = totalAmount;

            dbContext.Orders.Add(order);
            dbContext.SaveChanges();

            return order;
        }

        public Order UpdateOrder(int orderId, OrderDto orderDto)
        {
            Order order = dbContext.Orders.Find(orderId);

            if (order == null)
            {
                throw new Exception($"Ordine non trovato: {orderId}");
            }

            Customer customer = dbContext.Customers.Find(orderDto.CustomerId);

            if (customer == null)
            {
                throw new Exception($"Cliente non trovato: {orderDto.CustomerId}");
            }

            order.CustomerId = customer.Id;
            order.ShippingAddress = orderDto.ShippingAddress;
            order.OrderEmail = customer.Email;
            order.OrderStatus = orderDto.OrderStatus;


            List<OrderDetail> orderDetails = dbContext.OrderDetails.Where(x => x.OrderId == orderId).ToList();
            foreach (var orderDetail in orderDetails)
            {
                dbContext.OrderDetails.Remove(orderDetail);
            }

            decimal totalAmount = 0;

            foreach (var productDto in orderDto.Products)
            {
                var product = dbContext.Products.Find(productDto.ProductId);
                if (product == null)
                {
                    throw new Exception($"Prodotto non trovato: {productDto.ProductId}");
                }

                var orderDetail = new OrderDetail
                {
                    ProductId = productDto.ProductId,
                };

                order.OrderDetails.Add(orderDetail);

                totalAmount += (decimal)(product.Price * Convert.ToDecimal(productDto.Quantity));
            }

            order.Amount = totalAmount;

            dbContext.SaveChanges();

            return order;
        }
    }
}

File: ClientApp/src/app/services/orders.service.ts
import { Inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { OrderTable } from '../models/table-models';
import { Order, Product } from '../models/entity-models';

interface OrderDTO {
  customerId?: number,
  shippingAddress?: string,
  orderAddress?: string,
  orderEmail?: string,
  orderStatus?: string,
  products?: any[]
}

@Injectable({
  providedIn: 'root'
})
export class OrdersService {
  constructor(@Inject('BASE_URL') private baseUrl: string, private httpClient: HttpClient) {
    this.baseUrl = baseUrl + 'api/orders';
  }

  async addOrder(order: Order, productId: number, customerId: number): Promise<void> {
    let createOrderDTO: OrderDTO = {
      customerId,
      shippingAddress: order.shippingAddress,
      "orderAddress": "",
      "orderEmail": "",
      orderStatus: order.orderStatus,
      products: [
        {
          productId: productId,
          quantity: 1
        }
      ]
    }

    try {
      await this.httpClient.post(`${this.baseUrl}/createorder`, createOrderDTO).toPromise();
    } catch (err) {
      console.log(err);
      throw err;
    }
  }

  async updateOrder(order: Order, productId: number, customerId: number): Promise<void> {
    let createOrderDTO: OrderDTO = {
      customerId,
      shippingAddress: order.shippingAddress,
      "orderAddress": "",
      "orderEmail": "",
      orderStatus: order.orderStatus,
      products: [
        {
          productId: productId,
          quantity: 1
        }
      ]
    }

    try {
      await this.httpClient.post(`${this.baseUrl}/updateorder/${order.id}`, createOrderDTO).toPromise();
    } catch (err) {
      console.log(err);
      throw err;
    }
  }

  async getCustomerOder(orderId: number): Promise<any> {
    try {
      return this.httpClient.get<any>(`${this.baseUrl}/ordercustomer/${orderId}`).toPromise();
    } catch (err) {
      console.log(err);
      throw err;
    }
  }

  async getProductsOrder(orderId: number): Promise<any> {
    try {
      return this.httpClient.get<any>(`${this.baseUrl}/geproductsdorder/${orderId}`).toPromise();
    } catch (err) {
      console.log(err);
      throw err;
    }
  }

  async getPaginatedOrders(currentPage: number, pageSize: number, orderBy: string, orderDirection: string): Promise<OrderTable> {
    try {
      return this.httpClient.get<any>(`${this.baseUrl}/getpaginatedorders`, {
        params: {
          orderBy,
          orderDirection,
          currentPage: currentPage.toString(),
          pageSize: pageSize.toString()
        }
      }).toPromise();
    } catch (err) {
      console.log(err);
      throw err;
    }
  }
}
